File: path-sum/path-sum.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
// class Solution {
// public:
//     bool hasPathSum(TreeNode* root, int sum) {
//         if(root==NULL)
//             return false;
//         if(sum==0)
//             return true;
//         hasPathSum(root->left, sum-=root->val);
//         sum+=root->val;
//         hasPathSum(root->right, sum-=root->val);
//         sum+=root->val;
//         return false;
//     }
// };

class Solution {
public:
    bool hasPathSum(TreeNode* root, int sum) {
        if(root==NULL)
            return false;                                         
        sum=sum-root->val;                                           
        if(sum==0&&!root->left&&!root->right)
            return true; 
        bool left_flag=hasPathSum(root->left,sum);
        bool right_flag=hasPathSum(root->right,sum);
        return left_flag||right_flag;
    }
};

File: 112-path-sum/path-sum.py
class Solution:
    def solve(self, node, sum, targetSum):
        if node is None:
            return False
        if node.left is None and node.right is None:
            return sum+node.val == targetSum
        left = True
        right = True
        left = self.solve(node.left , sum+node.val, targetSum)
        right = self.solve(node.right, sum+node.val, targetSum)
        return left or right

    def hasPathSum(self, root: Optional[TreeNode], targetSum: int) -> bool:
        if not root:
            return False
        return self.solve(root, 0, targetSum)